crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+                            const OSSL_CMP_MSG *msg, X509 *cert)
+{
+    EVP_MD_CTX *ctx = NULL;
+    CMP_PROTECTEDPART prot_part;
+    int err = 1;
+    int digest_nid, pk_nid;
+    EVP_MD *digest = NULL;
+    EVP_PKEY *pubkey = NULL;
+    int l;
+    size_t prot_part_der_len = 0;
+    unsigned char *prot_part_der = NULL;
+
+    if (!ossl_assert(cmp_ctx != NULL && msg != NULL && cert != NULL))
+        return 0;
+
+    /* verify that keyUsage, if present, contains digitalSignature */
+    if (!cmp_ctx->ignore_keyusage
+            && (X509_get_key_usage(cert) & X509v3_KU_DIGITAL_SIGNATURE) == 0) {
+        CMPerr(0, CMP_R_MISSING_KEY_USAGE_DIGITALSIGNATURE);
+        err = 2;
+        goto cert_err;
+    }
+
+    pubkey = X509_get_pubkey(cert);
+    if (pubkey == NULL) {
+        CMPerr(0, CMP_R_FAILED_EXTRACTING_PUBKEY);
+        err = 2;
+        goto cert_err;
+    }
+
+    /* create the DER representation of protected part */
+    prot_part.header = msg->header;
+    prot_part.body = msg->body;
+
+    l = i2d_CMP_PROTECTEDPART(&prot_part, &prot_part_der);
+    if (l < 0 || prot_part_der == NULL)
+        goto end;
+    prot_part_der_len = (size_t) l;
+
+    /* verify signature of protected part */
+    if (!OBJ_find_sigid_algs(OBJ_obj2nid(msg->header->protectionAlg->algorithm),
+                             &digest_nid, &pk_nid)
+            || digest_nid == NID_undef || pk_nid == NID_undef
+            || (digest = (EVP_MD *)EVP_get_digestbynid(digest_nid)) == NULL) {
+        CMPerr(0, CMP_R_ALGORITHM_NOT_SUPPORTED);
+        err = 2;
+        goto end;
+    }
+
+    /* check msg->header->protectionAlg is consistent with public key type */
+    if (EVP_PKEY_type(pk_nid) != EVP_PKEY_base_id(pubkey)) {
+        CMPerr(0, CMP_R_WRONG_ALGORITHM_OID);
+        err = 2;
+        goto end;
+    }
+
+    if ((ctx = EVP_MD_CTX_new()) == NULL)
+        goto end;
+    err = (EVP_VerifyInit_ex(ctx, digest, NULL)
+               && EVP_VerifyUpdate(ctx, prot_part_der, prot_part_der_len)
+               && EVP_VerifyFinal(ctx, msg->protection->data,
+                                  msg->protection->length, pubkey) == 1)
+        ? 0 : 2;
+
+ end:
+    EVP_MD_CTX_free(ctx);
+    OPENSSL_free(prot_part_der);
+    EVP_PKEY_free(pubkey);
+
+ cert_err:
+    if (err == 2) { /* print diagnostics on cert verification error */
+        X509_STORE *ts = cmp_ctx->trusted;
+        X509_STORE_CTX *csc = X509_STORE_CTX_new();
+        X509_STORE_CTX_verify_cb verify_cb = ts == NULL ? NULL
+                                             : X509_STORE_get_verify_cb(ts);
+
+        if (csc != NULL && verify_cb != NULL /* implies ts != NULL */
+                && X509_STORE_CTX_init(csc, ts, NULL, NULL)) {
+            X509_STORE_CTX_set_current_cert(csc, cert);
+            X509_STORE_CTX_set_error_depth(csc, -1);
+            X509_STORE_CTX_set_error(csc, X509_V_ERR_UNSPECIFIED);
+            (void)(*verify_cb)(0, csc);
```

Comment:
What is going on here? Why are we invoking the verify_cb directly???

Reply:
The verification error callback `verify_cb` is called here after the CMP code detected that things have gone wrong checking the signature of a CMP message.
Note that these are not generic cert verification errors - these of course are detected by the x509 code, which then also calls `verify_cb` (for these types of errors). We make use of this in `OSSL_CMP_validate_cert_path)`.

Reply:
Good that you brought this up.
I've just realized that the label name `cert_err` and the comment was misleading here; I'll change that and move the sig err handling into an aux function.

Reply:
`verify_cb` is about verifying whether or not a certificate is valid and we trust it. This is an entirely different operation to validating a signature with the public key contained within a certificate. In the case of a valid certificate but an invalid signature I would expect the certificate "verify" stage to complete successfully and report no problems. The "verify" step is looking internally at the certificate and its associated chain. The signature is held externally to the certificate.

Using `verify_cb` here is confusing these two different operations. IMO this code should be removed.

Reply:
Good point.
I've removed that strange use of `verify_cb`, which - although I had to add some compensating diagnostic code for the case of CMP message signature check failure - simplified the overall code.
You'll see the new version after the next push by @Akretsch.